### ISSUE TITLE
Support custom shell when invoking build section

### ISSUE THREAD

**Issue by fracting** [drone/drone#1409]
I'd like to use non default shell to build projects, is that an valid feature request? If not, could you point out where is the right place to change so I can use this specific feature in my custom fork?

The reason I need a custom shell is, I'm developing a Drone based CI for Windows software, it use a shell wrapper to invoke Wine+Cygwin/MSYS2 in order to build Windows software.
The custom shell is called msys2-shell:
https://github.com/fracting/msys32-docker/blob/master/msys2-shell

Usage:

$ echo ls | msys2-shell
`# This will run `ls` inside MSYS2 shell on Wine in Linux`

$ cat /path/to/build.sh | msys2-shell
`# This will run /path/to/build.sh inside MSYS2 shell on Wine in Linux`

I built a docker image in DockerHub with Wine and my msys2-shell pre-installed: https://hub.docker.com/r/fracting/msys32-docker/builds/

Currently I explicitly invoke msys2-shell in my build script, so that it can build and test Windows software in drone.

My future plan is hack Drone to support msys2-shell as a custom shell, so Windows users can simply use the following syntax to build Windows software in a MSYS2 shell environment:

```
build:
  image: msys32-docker:latest
  emulate_os: WINE_MSYS32 # or WINE_MSYS64, WINE_MINGW32, WINE_MINGW64, WINE_CYGWIN32, WINE_CYGWIN64
  commands:
    - ./configure
    - make
    - make test
```

or 

```
build:
  image: msys32-docker:latest
  custom_shell: /usr/bin/msys32-shell # or /usr/bin/msys64-shell, /usr/bin/mingw32-shell, /usr/bin/mingw64-shell, /usr/bin/cygwin32-shell, /usr/bin/cygwin64-shell
  commands:
    - ./configure
    - make
    - make test
```

Any comments are great appreciated, thanks for the great project!

**Comment by tboerger** [comment]
Maybe we can add something like an optional template for generating the script that get's executed?

**Comment by fracting** [comment]
Hi @tboerger, thanks a lot for comment.

I'm not a go developer, but based on pure guess I think the below code is related:

```
// encode is a helper function that base64 encodes
// a shell command (or entire script)
func encode(script []byte) string {
        encoded := base64.StdEncoding.EncodeToString(script)
        return fmt.Sprintf("echo %s | base64 -d | /bin/sh", encoded)
}
```

For me I think replace `base64 -d | /bin/sh` with `base64 -d | /usr/bin/msys32-shell` seems to be the easiest solution. To make it extendable, we can read the custom shell path from .yml and insert it to that base64 command line.

Maybe something like below:

```
// encode is a helper function that base64 encodes
// a shell command (or entire script)
func encode(script []byte) string {
        encoded := base64.StdEncoding.EncodeToString(script)
        return fmt.Sprintf("echo %s | base64 -d", encoded)  // don't pipe to shell here 
}
```

```
func Encode(w *plugin.Workspace, c *dockerclient.ContainerConfig, n *parser.DockerNode) {
{
    c.Cmd = []string{encode(buf.Bytes())} + " | /path/to/custom_shell"
} 
```

Does that sounds fine?

I have no preference between one or another solution, if an optional template works, it would be fine for me as well. Thank you!

**Comment by bradrydzewski** [comment]
What about replacing `/bin/sh` inside your build container with a custom implementation that bootstraps your environment and shell?

**Comment by fracting** [comment]
@bradrydzewski I'm not sure this will be an general solution, what if /bin/custom_shell indirectly calls /bin/sh at some point? If /bin/sh is replaced to /bin/custom_shell and /bin/custom_shell indirectly calls /bin/sh, than there would be an infinite loop.

**Comment by bradrydzewski** [comment]
If you need a separate shell as part of your build process you can always do this:

```
build:
  image: foo
  commands:
    - /bin/dash build.sh
```

**Comment by fracting** [comment]
@bradrydzewski Thanks for comment. Sure, that's similar to the solution what I'm currently using. It works, it is just not *sexy* :) My purpose is to provide a "native" experience to users, so that user can freely use Windows cmd / Windows Cygwin / Windows msys2 / etc in their .yml file and run on Drone.

But if that's not a valid feature request, please feel free to close this issue, I still appreciate your great work ;)

**Comment by bradrydzewski** [comment]
@fracting at this time I don't think it is a customization we would add, however, it is something I would definitely consider in the future if there is enough demand. Let's keep the issue open and address again once more people are using the WineCI and can give us more real world usage feedback.

**Comment by fracting** [comment]
@bradrydzewski make sense, thank you!

**Comment by bradrydzewski** [comment]
this is required for windows (powershell vs cmd) and has therefore been implemented and will be available in the 0.9 release.
